src/libstd/sys/windows/stdio.rs

```
@@ -1,140 +1,226 @@
     pub fn new() -> io::Result<Stdin> {
-        Ok(Stdin {
-            utf8: Mutex::new(Cursor::new(Vec::new())),
-        })
+        Ok(Stdin { high_surrogate: Cell::new(0) })
     }
 
     pub fn read(&self, buf: &mut [u8]) -> io::Result<usize> {
-        let handle = match get(c::STD_INPUT_HANDLE)? {
-            Output::Console(c) => c,
-            Output::Pipe(p) => {
-                let handle = Handle::new(p);
-                let ret = handle.read(buf);
-                handle.into_raw();
-                return ret
-            }
+        let handle = get_handle(c::STD_INPUT_HANDLE)?;
+        if !is_console(handle) {
+            let handle = Handle::new(handle);
+            let ret = handle.read(buf);
+            handle.into_raw(); // Don't close the handle
+            return ret;
+        }
+
+        if buf.len() == 0 {
+            return Ok(0);
+        } else if buf.len() < 4 {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput,
+                        "Windows stdin in console mode does not support a buffer too small to; \
+                        guarantee holding one arbitrary UTF-8 character (4 bytes)"))
+        }
+
+        let mut utf16_buf = [0u16; MAX_BUFFER_SIZE / 2];
+        // In the worst case, an UTF-8 string can take 3 bytes for every `u16` of an UTF-16. So
+        // we can read at most a third of `buf.len()` chars and uphold the guarantee no data gets
+        // lost.
+        let amount = cmp::min(buf.len() / 3, utf16_buf.len());
+        let read = self.read_u16s_fixup_surrogates(handle, &mut utf16_buf, amount)?;
+        let utf16 = &utf16_buf[..read];
+
+        // FIXME: it would be nice if we could directly decode into the buffer instead of doing an
+        //        allocation.
+        let data = match String::from_utf16(&utf16) {
+            Ok(utf8) => utf8.into_bytes(),
+            Err(..) => {
+                // We can't really do any better than forget all data and return an error.
```

Two other possible ideas here:

* We could infinitely return errors
* We could return an error and then report EOF

I'm not really sure what the best handling mechanism here is, but dropping the data with an error doesn't seem entirely unreasonable.